ci/request-reviews/verify-base-branch.sh

```diff
@@ -0,0 +1,96 @@
+readarray -t devBranches < "$tmp/dev-branches"
+
+if [[ "$baseRepo" == "$prRepo" ]] && elementIn "$prBranch" "${devBranches[@]}"; then
+    log "This PR merges $prBranch into $baseBranch, no commit check necessary"
+    exit 0
+fi
+
+# The current merge base of the PR
+prMergeBase=$(git -C "$localRepo" merge-base "$baseBranch" "$headRef")
+log "The PR's merge base with the base branch $baseBranch is $prMergeBase"
+
+# This is purely for debugging
+git -C "$localRepo" rev-list --reverse "$baseBranch".."$headRef" > "$tmp/pr-commits"
+log "The PR includes these $(wc -l < "$tmp/pr-commits") commits:"
+cat <"$tmp/pr-commits" >&2
+
+for testBranch in "${devBranches[@]}"; do
+
+    if [[ -z "$(git -C "$localRepo" rev-list -1 --since="1 month ago" "$testBranch")" ]]; then
+        log "Not checking $testBranch, was inactive for the last month"
+        continue
+    fi
+    log "Checking if commits from $testBranch are included in the PR"
+
+    # We need to check for any commits that are in the PR which are also in the test branch.
+    # We could check each commit from the PR individually, but that's unnecessarily slow.
+    #
+    # This does _almost_ what we want: `git rev-list --count headRef testBranch ^baseBranch`,
+    # except that it includes commits that are reachable from _either_ headRef or testBranch,
+    # instead of restricting it to ones reachable by both
+
+    # Easily fixable though, because we can use `git merge-base testBranch headRef`
+    # to get the least common ancestor (aka merge base) commit reachable by both.
+    # If the branch being tested is indeed the right base branch,
+    # this is then also the commit from that branch that the PR is based on top of.
+    testMergeBase=$(git -C "$localRepo" merge-base "$testBranch" "$headRef")
+
+    # And then use the `git rev-list --count`, but replacing the non-working
+    # `headRef testBranch` with the merge base of the two.
+    extraCommits=$(git -C "$localRepo" rev-list --count "$testMergeBase" ^"$baseBranch")
+
+    if (( extraCommits != 0 )); then
+        log -en "\e[31m"
+        log "The PR's base branch is set to $baseBranch, but $extraCommits commits from the $testBranch branch are included."
+        log "Make sure you know the right base branch for your changes: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#branch-conventions, then:"
+        log "- If the changes should go to the $testBranch branch instead, change the base branch accordingly:"
+        log "  https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request"
+        log "- If the changes should go to the $baseBranch branch, rebase your PR onto merge base with the $testBranch branch:"
+        log "  git rebase --onto $prMergeBase $testMergeBase && git push --force-with-lease"
+        log -en "\e[0m"
+        exit 1
+    fi
```

Comment:
Could we perhaps comment this on the PR instead, for better visibility? The user will probably get a "workflow failed" email, but meh.

(I assume you decided on this for a reason, but just in case you didn't think of it, figured I'd bring it up.)

Reply:
I did think of it briefly. It does introduce more side effects: Comments could end up being posted multiple times for subsequent pushes, as each triggers a CI run. While requesting reviews is also a side-effect, it's an idempotent one, so it doesn't matter if it's done many times.

However thinking about this more deeply, I think the benefit of guiding newcomers in how to handle such scary situation smoothly outweighs the downside of an occasional double comment (if it even ever occurs), so let's do this :)

Reply:
> Comments could end up being posted multiple times for subsequent pushes

We could do a GH API query and quickly grep for our (the app's) user ID, but now I'm trying to think if there'd be a situation where the message could be different per screw up down the line (e.g. user thinks they set to the right base branch, but then they screw up the rebase). Right now, this leads to the same comment, but what if we do some magic to get more intelligent remediation commands at some point? All's this to say that the double commenting may be a good thing, if my explanation makes sense.

Reply:
There's actually some helpful reusable actions to help with this, and I've used this before [here](https://github.com/NixOS/nixfmt/blob/d0c0cda4fe6b7ae2b9aa37f3ee15e6412573eb01/.github/workflows/main.yml#L39-L94) ([in action](https://github.com/NixOS/nixfmt/pull/254#issuecomment-2356858761)).

But yeah agreed, I think it's fine to just not worry about the double comment for now. We can revisit if it becomes a problem.

Reply:
Yup, sounds good.

Reply:
(leaving this unresolved until I implement this)

Reply:
Done and tested! https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/21